How many late days are you using for this assignment? 0

In [15]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from scipy import stats
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [16]:
# Load the given datasets
df_c = pd.read_csv('chronic_kidney_disease_categorical.csv')
df_n = pd.read_csv('chronic_kidney_disease_numerical.csv')

In [17]:
# Print the data
df_c

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


In [18]:
df_n

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The label is 'class' and it is a binary variable stating whether or not someone has or does not have chronic kidney disease. The features in the given datasets are categorical and numerical. Examples of categorical variables are specific gravity (sg), al (alnumin), and sugar (su). Examples of numerical variables are blood pressure (bp), age, and white cell blood count (wbcc).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [19]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print('Duplicates?:', df_n['unique_id'].nunique() == df_n.shape[0]) # if false, then there are duplicates
print('Num duplicates in numerical dataset:', df_n.shape[0] - df_n['unique_id'].nunique())

# Drop these duplicate rows
df_n = df_n.drop_duplicates(subset=['unique_id'])
print('No more duplicates?:', df_n['unique_id'].nunique() == df_n.shape[0])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print('Duplicates?:', df_c['unique_id'].nunique() == df_c.shape[0]) # if false, then there are duplicates
print('Num duplicates in numerical dataset:', df_c.shape[0] - df_c['unique_id'].nunique())

# Drop these duplicate rows
df_c = df_c.drop_duplicates(subset=['unique_id'])
print('No more duplicates?:', df_c['unique_id'].nunique() == df_c.shape[0])



Duplicates?: False
Num duplicates in numerical dataset: 30
No more duplicates?: True
Duplicates?: False
Num duplicates in numerical dataset: 30
No more duplicates?: True


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: We can see from printing the shape that the numerical dataset does not have all the unique ids present in the categorical dataset. The rows that are missing from one of the datasets should be discarded since 10 rows is not a substantial portion of the data and they don't have enough features to be considered useful to train a prediction model.

In [20]:
print(df_c.shape)
print(df_n.shape)

(400, 14)
(390, 12)


In [21]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df = pd.merge(df_c, df_n, on='unique_id')
#Print the combined dataset
df

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,...,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,...,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,...,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
387,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,...,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8
388,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [22]:
# Calculate the percentage of rows that contain atleast one missing value
# Print %
print((df.shape[0] - df.dropna(axis=0,inplace=False).shape[0]) / df.shape[0])

# Drop these rows from the dataset
df.dropna(axis=0,inplace=True)
# Print the Dataset
df

0.6076923076923076


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,...,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
382,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
384,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,...,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.4: Sort the dataset according to the Labels

In [23]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df.sort_values(by=['Target'], inplace=True)
df.reset_index(inplace=True,drop=True)
# Print the dataset
df

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,534245,2.0,0.0,normal,abnormal,present,notpresent,no,no,no,...,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7
1,349892,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1
2,343710,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2
3,102677,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1
4,514721,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0
149,530138,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0
150,261137,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2
151,322882,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5


In [24]:
df['Target']

0         ckd
1         ckd
2         ckd
3         ckd
4         ckd
        ...  
148    notckd
149    notckd
150    notckd
151    notckd
152    notckd
Name: Target, Length: 153, dtype: object

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [25]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df = pd.get_dummies(df, columns=categorical_columns, dtype=int)

# Print the dataset
df

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,534245,2.0,0.0,45.0,70.0,117.0,52.0,2.2,136.0,3.8,...,1,0,1,0,1,0,1,0,1,0
1,349892,1.0,3.0,59.0,70.0,424.0,55.0,1.7,138.0,4.5,...,0,1,1,0,1,0,1,0,1,0
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
3,102677,3.0,0.0,59.0,70.0,76.0,186.0,15.0,135.0,7.6,...,1,0,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


In [26]:
df.columns

Index(['unique_id', 'al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot',
       'hemo', 'pcv', 'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: In this solution and example, we would assign each of the 5 values in the original categorical column with a number from 1-5.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [27]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
df = df[np.abs(stats.zscore(df[numerical_columns])<3).all(axis=1)]
# Print the dataset
df

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,69.0,70.0,264.0,87.0,2.7,130.0,4.0,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,62.0,70.0,122.0,42.0,1.7,136.0,4.7,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,6.0,60.0,94.0,67.0,1.0,135.0,4.9,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [28]:
# Normalize the all Numerical Attributes in the dataset.
# use min-max normalization along each numerical feature
scaler = MinMaxScaler()
df.loc[:,numerical_columns] = scaler.fit_transform(df[numerical_columns])

# Print the dataset
df

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,0.714286,0.50,0.442060,0.901961,0.432099,0.615385,0.692308,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,0.649351,0.75,0.253219,0.633987,0.777778,0.512821,0.589744,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,0.818182,0.25,0.832618,0.503268,0.283951,0.487179,0.384615,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,0.727273,0.25,0.223176,0.209150,0.160494,0.641026,0.564103,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.615385,0.615385,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,0.285714,0.00,0.038627,0.261438,0.012346,0.871795,0.641026,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,0.220779,0.50,0.175966,0.156863,0.086420,0.871795,0.384615,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,0.376623,0.00,0.150215,0.189542,0.012346,0.615385,0.358974,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,0.350649,0.50,0.248927,0.183007,0.024691,0.615385,0.358974,...,1,0,1,0,1,0,1,0,0,1


In [16]:
df.columns

Index(['unique_id', 'al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot',
       'hemo', 'pcv', 'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: We don't need Target_notckd because it gives the same information as Target_ckd. Similarly, we also don't need the complements of the other binary dummy variables like rbc_normal, pc_normal, pcc_notpresent, ba_notpresent, htn_no, dm_no, cad_no, appet_good, pe_no, and ane_no.

In [29]:
#Remove that column
to_remove = ['Target_notckd', 'rbc_normal', 'pc_normal', 'pcc_notpresent', 'ba_notpresent', 'htn_no', 'dm_no', 'cad_no', 'appet_good', 'pe_no', 'ane_no']
df = df.drop(columns=to_remove)

# Print the dataset
df

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,pc_abnormal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_ckd
2,343710,2.0,0.0,0.714286,0.50,0.442060,0.901961,0.432099,0.615385,0.692308,...,1,0,0,1,1,1,1,1,1,1
4,514721,2.0,0.0,0.649351,0.75,0.253219,0.633987,0.777778,0.512821,0.589744,...,1,0,0,1,0,0,0,0,0,1
7,124304,3.0,4.0,0.818182,0.25,0.832618,0.503268,0.283951,0.487179,0.384615,...,1,0,0,1,1,1,0,1,0,1
8,980291,3.0,0.0,0.727273,0.25,0.223176,0.209150,0.160494,0.641026,0.564103,...,1,0,0,1,1,0,0,0,0,1
10,615697,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.615385,0.615385,...,1,0,1,0,0,0,1,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,0.285714,0.00,0.038627,0.261438,0.012346,0.871795,0.641026,...,0,0,0,0,0,0,0,0,0,0
149,530138,0.0,0.0,0.220779,0.50,0.175966,0.156863,0.086420,0.871795,0.384615,...,0,0,0,0,0,0,0,0,0,0
150,261137,0.0,0.0,0.376623,0.00,0.150215,0.189542,0.012346,0.615385,0.358974,...,0,0,0,0,0,0,0,0,0,0
151,322882,0.0,0.0,0.350649,0.50,0.248927,0.183007,0.024691,0.615385,0.358974,...,0,0,0,0,0,0,0,0,0,0


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [30]:
# Export the dataframe to a new csv file
df.to_csv('cleaned_data.csv', index=False)
# Export the dataframe to a new json file
df.to_json('cleaned_data.json', index=False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell below

Paste here:

### Paste the markdown table here

| unique_id | al    | su    | age          | bp    | bgr          | bu          | sc          | sod          | pot          | hemo         | pcv          | wbcc         | rbcc         | rbc_abnormal | pc_abnormal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_ckd |
|-----------|-------|-------|--------------|-------|--------------|-------------|-------------|--------------|--------------|---------------|--------------|--------------|--------------|---------------|-------------|-------------|------------|---------|--------|---------|------------|--------|---------|------------|
| 343710    | 2.000 | 0.000 | 0.71429      | 0.500 | 0.44206      | 0.90196     | 0.43210     | 0.61538      | 0.69231      | 0.17213       | 0.21053      | 0.39516      | 0.15385      | 1             | 1           | 0           | 0          | 1       | 1      | 1       | 1          | 1      | 1       | 1          |
| 514721    | 2.000 | 0.000 | 0.64935      | 0.750 | 0.25322      | 0.63399     | 0.77778     | 0.51282      | 0.58974      | 0.28689       | 0.34211      | 0.16935      | 0.20513      | 1             | 1           | 0           | 0          | 1       | 0      | 0       | 0          | 0      | 0       | 1          |
| 124304    | 3.000 | 4.000 | 0.81818      | 0.250 | 0.83262      | 0.50327     | 0.28395     | 0.48718      | 0.38462      | 0.56557       | 0.55263      | 0.42742      | 0.38462      | 0             | 1           | 0           | 0          | 1       | 1      | 1       | 0          | 1      | 0       | 1          |
| 980291    | 3.000 | 0.000 | 0.72727      | 0.250 | 0.22318      | 0.20915     | 0.16049     | 0.64103      | 0.56410      | 0.57377       | 0.60526      | 0.29032      | 0.33333      | 0             | 1           | 0           | 0          | 1       | 1      | 0       | 0          | 0      | 0       | 1          |
| 615697    | 4.000 | 0.000 | 0.00000      | 0.000 | 0.10300      | 0.37255     | 0.07407     | 0.61538      | 0.61538      | 0.35246       | 0.36842      | 1.00000      | 0.56410      | 1             | 1           | 0           | 1          | 0       | 0      | 0       | 1          | 0      | 0       | 1          |
| 955830    | 3.000 | 1.000 | 0.63636      | 0.500 | 0.61803      | 0.41176     | 0.43210     | 0.66667      | 0.61538      | 0.43443       | 0.47368      | 0.25000      | 0.28205      | 0             | 1           | 1           | 1          | 1       | 1      | 0       | 0          | 1      | 0       | 1          |
| 899916    | 3.000 | 0.000 | 1.00000      | 0.250 | 0.13734      | 0.32680     | 0.27160     | 0.10256      | 0.82051      | 0.25410       | 0.26316      | 0.68548      | 0.12821      | 0             | 0           | 0           | 0          | 1       | 0      | 0       | 1          | 0      | 1       | 1          |
| 546225    | 1.000 | 0.000 | 0.51948      | 0.000 | 0.39914      | 0.53595     | 0.35802     | 0.76923      | 0.38462      | 0.34426       | 0.31579      | 0.83065      | 0.15385      | 0             | 0           | 0           | 0          | 1       | 1      | 0       | 0          | 0      | 0       | 1          |
| 475200    | 4.000 | 2.000 | 0.75325      | 1.000 | 0.39914      | 0.28758     | 0.83951     | 0.74359      | 0.53846      | 0.18852       | 0.26316      | 0.25806      | 0.20513      | 1             | 1           | 0           | 1          | 1       | 1      | 0       | 0          | 1      | 0       | 1          |
| 234299    | 4.000 | 0.000 | 0.54545      | 0.500 | 0.10730      | 1.00000     | 0.90123     | 0.64103      | 0.33333      | 0.34426       | 0.42105      | 0.20968      | 0.20513      | 0             | 1           | 0           | 0          | 1       | 0      | 0       | 0          | 0      | 1       | 1          |
| 995177    | 4.000 | 3.000 | 0.81818      | 0.250 | 0.61803      | 0.56209     | 0.72840     | 0.23077      | 0.35897      | 0.31148       | 0.31579      | 0.58065      | 0.17949      | 0             | 1           | 1           | 1          | 1       | 1      | 1       | 0          | 1      | 1       | 1          |
| 137148    | 3.000 | 2.000 | 0.87013      | 1.000 | 0.96567      | 0.52288     | 0.64198     | 0.74359      | 0.10256      | 0.29508       | 0.36842      | 0.21774      | 0.15385      | 1             | 1           | 1           | 0          | 1       | 1      | 1       | 1          | 0      | 0       | 1          |
| 474407    | 4.000 | 0.000 | 0.19481      | 0.750 | 0.15880      | 0.19608     | 0.16049     | 0.35897      | 0.25641      | 0.22131       | 0.18421      | 0.65323      | 0.33333      | 0             | 1           | 1           | 1          | 0       | 0      | 0       | 0          | 0      | 1       | 1          |
| 923613    | 4.000 | 1.000 | 0.75325      | 0.000 | 0.72532      | 0.31373     | 0.48148     | 0.66667      | 0.74359      | 0.31967       | 0.34211      | 0.25806      | 0.20513      | 1             | 1           | 0           | 1          | 1       | 1      | 0       | 1          | 1      | 0       | 1          |
| 197452    | 4.000 | 1.000 | 0.64935      | 0.250 | 0.60086      | 0.10458     | 0.16049     | 0.64103      | 0.33333      | 0.86066       | 0.94737      | 0.66129      | 0.76923      | 1             | 0           | 0           | 0          | 0       | 0      | 0       | 0          | 0      | 0       | 1          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

You could be more detailed in your prompts when prompting the LLM and specify explicitly rules for converting data such as data format and data type.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [31]:
# Code to rename all the columns in the dataset
# Create a mapping of current column names to preferred names
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Apply the renaming to the DataFrame
df.rename(columns=column_mapping, inplace=True)

# Print the DataFrame to check the updated column names
df.head()

,unique_id,albumin,sugar,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,...,pc_abnormal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_ckd
2,343710,2.0,0.0,0.714286,0.50,0.442060,0.901961,0.432099,0.615385,0.692308,...,1,0,0,1,1,1,1,1,1,1
4,514721,2.0,0.0,0.649351,0.75,0.253219,0.633987,0.777778,0.512821,0.589744,...,1,0,0,1,0,0,0,0,0,1
7,124304,3.0,4.0,0.818182,0.25,0.832618,0.503268,0.283951,0.487179,0.384615,...,1,0,0,1,1,1,0,1,0,1
8,980291,3.0,0.0,0.727273,0.25,0.223176,0.209150,0.160494,0.641026,0.564103,...,1,0,0,1,1,0,0,0,0,1
10,615697,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.615385,0.615385,...,1,0,1,0,0,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [32]:
# Converted SQL to Pandas code
# I changed Target to Target_ckd since it's the name of the column in the dataset
result = df.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
result

,Target_ckd,count
0,0,111
1,1,26


The above code groups the values of the column Target_ckd and displays how many are in each group.